File: Problem199/Solution.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
	var result []int
	populateRightSideView(root, &result, 0)
	return result
}

func populateRightSideView(root *TreeNode, result *[]int, level int) {
	if root == nil {
		return
	}
	if level == len(*result) {
		*result = append(*result, root.Val)
	}
	populateRightSideView(root.Right, result, level+1)
	populateRightSideView(root.Left, result, level+1)
}

File: Problem199/Solution.java
public class Solution
{
    public List<Integer> rightSideView(TreeNode root)
    {
        List<Integer> list = new ArrayList<Integer>();
        populateRightSideView(root, 0, list);
        return list;
    }

    private void populateRightSideView(TreeNode root, int level, List<Integer> list)
    {
        if (root == null)
        {
            return;
        }
        if (level == list.size())
        {
            list.add(root.val);
        }
        populateRightSideView(root.right, level + 1, list);
        populateRightSideView(root.left, level + 1, list);
    }
}